File: benchmarks/cross_language_overhead.ts
/**
 * Cross-Language Overhead Measurement - Node.js Binding
 *
 * This measures FFI overhead for Node.js bindings compared to the Rust baseline.
 *
 * Run with: npx tsx benchmarks/cross_language_overhead.ts
 */

const ITERATIONS = 100_000;
const WARMUP_ITERATIONS = 1_000;

interface BenchResult {
  name: string;
  iterations: number;
  total_ns: number;
  avg_ns: number;
  throughput: number;
}

function newBenchResult(
  name: string,
  iterations: number,
  totalNs: number
): BenchResult {
  const avgNs = totalNs / iterations;
  const throughput = 1_000_000_000 / avgNs;
  return {
    name,
    iterations,
    total_ns: totalNs,
    avg_ns: avgNs,
    throughput,
  };
}

function printResult(r: BenchResult): void {
  console.log(
    `${r.name.padEnd(40)} ${r.iterations.toString().padStart(10)} iterations, ${r.avg_ns.toFixed(2).padStart(10)} ns/op, ${r.throughput.toFixed(0).padStart(12)} ops/sec`
  );
}

function warmup(f: () => void): void {
  for (let i = 0; i < WARMUP_ITERATIONS; i++) {
    f();
  }
}

function bench(name: string, iterations: number, f: () => void): BenchResult {
  warmup(f);

  const start = process.hrtime.bigint();
  for (let i = 0; i < iterations; i++) {
    f();
  }
  const elapsed = process.hrtime.bigint() - start;

  return newBenchResult(name, iterations, Number(elapsed));
}

// ============================================================================
// Geometry Types (Pure TypeScript implementation for baseline)
// ============================================================================

class Point {
  constructor(
    public x: number,
    public y: number
  ) {}

  transform(m: Matrix): Point {
    return new Point(
      this.x * m.a + this.y * m.c + m.e,
      this.x * m.b + this.y * m.d + m.f
    );
  }
}

class Rect {
  constructor(
    public x0: number,
    public y0: number,
    public x1: number,
    public y1: number
  ) {}

  contains(p: Point): boolean {
    return p.x >= this.x0 && p.x <= this.x1 && p.y >= this.y0 && p.y <= this.y1;
  }

  union(other: Rect): Rect {
    return new Rect(
      Math.min(this.x0, other.x0),
      Math.min(this.y0, other.y0),
      Math.max(this.x1, other.x1),
      Math.max(this.y1, other.y1)
    );
  }

  transform(m: Matrix): Rect {
    const p1x = this.x0 * m.a + this.y0 * m.c + m.e;
    const p1y = this.x0 * m.b + this.y0 * m.d + m.f;
    const p2x = this.x1 * m.a + this.y0 * m.c + m.e;
    const p2y = this.x1 * m.b + this.y0 * m.d + m.f;
    const p3x = this.x0 * m.a + this.y1 * m.c + m.e;
    const p3y = this.x0 * m.b + this.y1 * m.d + m.f;
    const p4x = this.x1 * m.a + this.y1 * m.c + m.e;
    const p4y = this.x1 * m.b + this.y1 * m.d + m.f;

    return new Rect(
      Math.min(p1x, p2x, p3x, p4x),
      Math.min(p1y, p2y, p3y, p4y),
      Math.max(p1x, p2x, p3x, p4x),
      Math.max(p1y, p2y, p3y, p4y)
    );
  }
}

class Matrix {
  constructor(
    public a: number,
    public b: number,
    public c: number,
    public d: number,
    public e: number,
    public f: number
  ) {}

  static identity(): Matrix {
    return new Matrix(1, 0, 0, 1, 0, 0);
  }

  static scale(sx: number, sy: number): Matrix {
    return new Matrix(sx, 0, 0, sy, 0, 0);
  }

  static rotate(degrees: number): Matrix {
    const rad = (degrees * Math.PI) / 180;
    const cos = Math.cos(rad);
    const sin = Math.sin(rad);
    return new Matrix(cos, sin, -sin, cos, 0, 0);
  }

  static translate(tx: number, ty: number): Matrix {
    return new Matrix(1, 0, 0, 1, tx, ty);
  }

  concat(other: Matrix): Matrix {
    return new Matrix(
      this.a * other.a + this.b * other.c,
      this.a * other.b + this.b * other.d,
      this.c * other.a + this.d * other.c,
      this.c * other.b + this.d * other.d,
      this.e * other.a + this.f * other.c + other.e,
      this.e * other.b + this.f * other.d + other.f
    );
  }

  invert(): Matrix | null {
    const det = this.a * this.d - this.b * this.c;
    if (Math.abs(det) < 1e-14) {
      return null;
    }
    const invDet = 1 / det;
    return new Matrix(
      this.d * invDet,
      -this.b * invDet,
      -this.c * invDet,
      this.a * invDet,
      (this.c * this.f - this.d * this.e) * invDet,
      (this.b * this.e - this.a * this.f) * invDet
    );
  }
}

// Buffer simulation
class Buffer {
  private data: Uint8Array;
  private length: number;

  constructor(capacity: number) {
    this.data = new Uint8Array(capacity);
    this.length = 0;
  }

  static fromSlice(data: Uint8Array): Buffer {
    const buf = new Buffer(data.length);
    buf.data.set(data);
    buf.length = data.length;
    return buf;
  }

  appendData(chunk: Uint8Array): void {
    if (this.length + chunk.length > this.data.length) {
      const newData = new Uint8Array(
        Math.max(this.data.length * 2, this.length + chunk.length)
      );
      newData.set(this.data);
      this.data = newData;
    }
    this.data.set(chunk, this.length);
    this.length += chunk.length;
  }
}

// ============================================================================
// Main Benchmark
// ============================================================================

function main(): void {
  console.log('=== Cross-Language Overhead Benchmark (Node.js) ===\n');
  console.log(`Iterations: ${ITERATIONS}`);
  console.log(`Warmup: ${WARMUP_ITERATIONS}`);
  console.log();

  const results: BenchResult[] = [];

  // ========================================================================
  // Geometry Operations
  // ========================================================================

  console.log('--- Geometry Operations ---\n');

  // Point creation
  results.push(
    bench('point_create', ITERATIONS, () => {
      new Point(10.0, 20.0);
    })
  );

  // Point transform
  const p = new Point(10.0, 20.0);
  let m = Matrix.scale(2.0, 2.0);
  results.push(
    bench('point_transform', ITERATIONS, () => {
      p.transform(m);
    })
  );

  // Rect creation
  results.push(
    bench('rect_create', ITERATIONS, () => {
      new Rect(0.0, 0.0, 100.0, 100.0);
    })
  );

  // Rect transform
  let r = new Rect(0.0, 0.0, 100.0, 100.0);
  m = Matrix.rotate(45.0);
  results.push(
    bench('rect_transform', ITERATIONS, () => {
      r.transform(m);
    })
  );

  // Rect contains point
  r = new Rect(0.0, 0.0, 100.0, 100.0);
  const testPoint = new Point(50.0, 50.0);
  results.push(
    bench('rect_contains_point', ITERATIONS, () => {
      r.contains(testPoint);
    })
  );

  // Matrix creation
  results.push(
    bench('matrix_identity', ITERATIONS, () => {
      Matrix.identity();
    })
  );

  results.push(
    bench('matrix_scale', ITERATIONS, () => {
      Matrix.scale(2.0, 2.0);
    })
  );

  results.push(
    bench('matrix_rotate', ITERATIONS, () => {
      Matrix.rotate(45.0);
    })
  );

  results.push(
    bench('matrix_translate', ITERATIONS, () => {
      Matrix.translate(100.0, 100.0);
    })
  );

  // Matrix concatenation
  const m1 = Matrix.scale(2.0, 2.0);
  const m2 = Matrix.rotate(45.0);
  results.push(
    bench('matrix_concat', ITERATIONS, () => {
      m1.concat(m2);
    })
  );

  // Matrix inversion
  m = Matrix.scale(2.0, 3.0);
  results.push(
    bench('matrix_invert', ITERATIONS, () => {
      m.invert();
    })
  );

  console.log();

  // ========================================================================
  // Buffer Operations
  // ========================================================================

  console.log('--- Buffer Operations ---\n');

  // Buffer creation
  results.push(
    bench('buffer_create_empty', ITERATIONS, () => {
      new Buffer(0);
    })
  );

  results.push(
    bench('buffer_create_1KB', ITERATIONS, () => {
      new Buffer(1024);
    })
  );

  // Buffer from data
  let data = new Uint8Array(256);
  results.push(
    bench('buffer_from_slice_256B', ITERATIONS, () => {
      Buffer.fromSlice(data);
    })
  );

  data = new Uint8Array(1024);
  results.push(
    bench('buffer_from_slice_1KB', ITERATIONS, () => {
      Buffer.fromSlice(data);
    })
  );

  // Buffer append
  const chunk = new Uint8Array(64);
  results.push(
    bench('buffer_append_64B', ITERATIONS / 10, () => {
      const buf = new Buffer(64);
      buf.appendData(chunk);
    })
  );

  console.log();

  // ========================================================================
  // Combined Operations
  // ========================================================================

  console.log('--- Combined Operations ---\n');

  // Simulate page rendering setup
  results.push(
    bench('page_render_setup', ITERATIONS / 10, () => {
      const dpi = 144.0;
      const scale = dpi / 72.0;
      const ctm = Matrix.scale(scale, scale);
      const pageBounds = new Rect(0.0, 0.0, 612.0, 792.0);
      pageBounds.transform(ctm);
    })
  );

  // Simulate text position calculation
  results.push(
    bench('text_position_calc', ITERATIONS, () => {
      const base = new Point(72.0, 700.0);
      const tm = Matrix.translate(0.0, -14.0);
      base.transform(tm);
    })
  );

  // Simulate bounding box calculation for 10 items
  results.push(
    bench('bbox_calc_10_items', ITERATIONS / 10, () => {
      let bbox = new Rect(1e30, 1e30, -1e30, -1e30);
      for (let i = 0; i < 10; i++) {
        const item = new Rect(i * 10, i * 10, 100 + i * 10, 20 + i * 10);
        bbox = bbox.union(item);
      }
    })
  );

  console.log();

  // ========================================================================
  // Summary
  // ========================================================================

  console.log('=== Summary ===\n');

  for (const result of results) {
    printResult(result);
  }

  console.log();

  // Output JSON for cross-language comparison
  console.log('--- JSON Output ---');
  console.log(JSON.stringify(results, null, 2));
}

main();

File: benchmarks/cross_language_overhead.py
#!/usr/bin/env python3
"""
Cross-Language Overhead Measurement - Python Binding

This measures FFI overhead for Python bindings compared to the Rust baseline.

Run with: python benchmarks/cross_language_overhead.py
"""

from __future__ import annotations

import json
import math
import time
from dataclasses import dataclass
from typing import Callable, List, Optional

ITERATIONS = 100_000
WARMUP_ITERATIONS = 1_000


@dataclass
class BenchResult:
    name: str
    iterations: int
    total_ns: int
    avg_ns: float
    throughput: float

    @staticmethod
    def create(name: str, iterations: int, total_ns: int) -> "BenchResult":
        avg_ns = total_ns / iterations
        throughput = 1_000_000_000.0 / avg_ns
        return BenchResult(
            name=name,
            iterations=iterations,
            total_ns=total_ns,
            avg_ns=avg_ns,
            throughput=throughput,
        )

    def print(self) -> None:
        print(
            f"{self.name:<40} {self.iterations:>10} iterations, "
            f"{self.avg_ns:>10.2f} ns/op, {self.throughput:>12.0f} ops/sec"
        )

    def to_dict(self) -> dict:
        return {
            "name": self.name,
            "iterations": self.iterations,
            "total_ns": self.total_ns,
            "avg_ns": self.avg_ns,
            "throughput": self.throughput,
        }


def warmup(f: Callable[[], None]) -> None:
    """Warmup function to trigger optimizations."""
    for _ in range(WARMUP_ITERATIONS):
        f()


def bench(name: str, iterations: int, f: Callable[[], None]) -> BenchResult:
    """Time a function."""
    warmup(f)

    start = time.perf_counter_ns()
    for _ in range(iterations):
        f()
    elapsed = time.perf_counter_ns() - start

    return BenchResult.create(name, iterations, elapsed)


# ============================================================================
# Geometry Types (Pure Python implementation for baseline)
# ============================================================================


class Point:
    """2D point."""

    __slots__ = ("x", "y")

    def __init__(self, x: float, y: float) -> None:
        self.x = float(x)
        self.y = float(y)

    def transform(self, m: "Matrix") -> "Point":
        return Point(
            self.x * m.a + self.y * m.c + m.e,
            self.x * m.b + self.y * m.d + m.f,
        )


class Rect:
    """Rectangle."""

    __slots__ = ("x0", "y0", "x1", "y1")

    def __init__(self, x0: float, y0: float, x1: float, y1: float) -> None:
        self.x0 = float(x0)
        self.y0 = float(y0)
        self.x1 = float(x1)
        self.y1 = float(y1)

    def contains(self, p: Point) -> bool:
        return self.x0 <= p.x <= self.x1 and self.y0 <= p.y <= self.y1

    def union(self, other: "Rect") -> "Rect":
        return Rect(
            min(self.x0, other.x0),
            min(self.y0, other.y0),
            max(self.x1, other.x1),
            max(self.y1, other.y1),
        )

    def transform(self, m: "Matrix") -> "Rect":
        # Transform all 4 corners
        p1x = self.x0 * m.a + self.y0 * m.c + m.e
        p1y = self.x0 * m.b + self.y0 * m.d + m.f
        p2x = self.x1 * m.a + self.y0 * m.c + m.e
        p2y = self.x1 * m.b + self.y0 * m.d + m.f
        p3x = self.x0 * m.a + self.y1 * m.c + m.e
        p3y = self.x0 * m.b + self.y1 * m.d + m.f
        p4x = self.x1 * m.a + self.y1 * m.c + m.e
        p4y = self.x1 * m.b + self.y1 * m.d + m.f

        return Rect(
            min(p1x, p2x, p3x, p4x),
            min(p1y, p2y, p3y, p4y),
            max(p1x, p2x, p3x, p4x),
            max(p1y, p2y, p3y, p4y),
        )


class Matrix:
    """2D transformation matrix."""

    __slots__ = ("a", "b", "c", "d", "e", "f")

    def __init__(
        self, a: float, b: float, c: float, d: float, e: float, f: float
    ) -> None:
        self.a = float(a)
        self.b = float(b)
        self.c = float(c)
        self.d = float(d)
        self.e = float(e)
        self.f = float(f)

    @staticmethod
    def identity() -> "Matrix":
        return Matrix(1.0, 0.0, 0.0, 1.0, 0.0, 0.0)

    @staticmethod
    def scale(sx: float, sy: float) -> "Matrix":
        return Matrix(float(sx), 0.0, 0.0, float(sy), 0.0, 0.0)

    @staticmethod
    def rotate(degrees: float) -> "Matrix":
        rad = math.radians(degrees)
        cos = math.cos(rad)
        sin = math.sin(rad)
        return Matrix(cos, sin, -sin, cos, 0.0, 0.0)

    @staticmethod
    def translate(tx: float, ty: float) -> "Matrix":
        return Matrix(1.0, 0.0, 0.0, 1.0, float(tx), float(ty))

    def concat(self, other: "Matrix") -> "Matrix":
        return Matrix(
            self.a * other.a + self.b * other.c,
            self.a * other.b + self.b * other.d,
            self.c * other.a + self.d * other.c,
            self.c * other.b + self.d * other.d,
            self.e * other.a + self.f * other.c + other.e,
            self.e * other.b + self.f * other.d + other.f,
        )

    def invert(self) -> Optional["Matrix"]:
        det = self.a * self.d - self.b * self.c
        if abs(det) < 1e-14:
            return None
        inv_det = 1.0 / det
        return Matrix(
            self.d * inv_det,
            -self.b * inv_det,
            -self.c * inv_det,
            self.a * inv_det,
            (self.c * self.f - self.d * self.e) * inv_det,
            (self.b * self.e - self.a * self.f) * inv_det,
        )


class Buffer:
    """Dynamic byte buffer."""

    __slots__ = ("_data",)

    def __init__(self, capacity: int) -> None:
        self._data = bytearray(capacity)

    @staticmethod
    def from_slice(data: bytes) -> "Buffer":
        buf = Buffer(len(data))
        buf._data[:] = data
        return buf

    def append_data(self, data: bytes) -> None:
        self._data.extend(data)


# ============================================================================
# Main Benchmark
# ============================================================================


def main() -> None:
    print("=== Cross-Language Overhead Benchmark (Python) ===\n")
    print(f"Iterations: {ITERATIONS}")
    print(f"Warmup: {WARMUP_ITERATIONS}")
    print()

    results: List[BenchResult] = []

    # ========================================================================
    # Geometry Operations
    # ========================================================================

    print("--- Geometry Operations ---\n")

    # Point creation
    results.append(bench("point_create", ITERATIONS, lambda: Point(10.0, 20.0)))

    # Point transform
    p = Point(10.0, 20.0)
    m = Matrix.scale(2.0, 2.0)
    results.append(bench("point_transform", ITERATIONS, lambda: p.transform(m)))

    # Rect creation
    results.append(
        bench("rect_create", ITERATIONS, lambda: Rect(0.0, 0.0, 100.0, 100.0))
    )

    # Rect transform
    r = Rect(0.0, 0.0, 100.0, 100.0)
    m = Matrix.rotate(45.0)
    results.append(bench("rect_transform", ITERATIONS, lambda: r.transform(m)))

    # Rect contains point
    r = Rect(0.0, 0.0, 100.0, 100.0)
    test_point = Point(50.0, 50.0)
    results.append(
        bench("rect_contains_point", ITERATIONS, lambda: r.contains(test_point))
    )

    # Matrix creation
    results.append(bench("matrix_identity", ITERATIONS, Matrix.identity))
    results.append(bench("matrix_scale", ITERATIONS, lambda: Matrix.scale(2.0, 2.0)))
    results.append(bench("matrix_rotate", ITERATIONS, lambda: Matrix.rotate(45.0)))
    results.append(
        bench("matrix_translate", ITERATIONS, lambda: Matrix.translate(100.0, 100.0))
    )

    # Matrix concatenation
    m1 = Matrix.scale(2.0, 2.0)
    m2 = Matrix.rotate(45.0)
    results.append(bench("matrix_concat", ITERATIONS, lambda: m1.concat(m2)))

    # Matrix inversion
    m = Matrix.scale(2.0, 3.0)
    results.append(bench("matrix_invert", ITERATIONS, lambda: m.invert()))

    print()

    # ========================================================================
    # Buffer Operations
    # ========================================================================

    print("--- Buffer Operations ---\n")

    # Buffer creation
    results.append(bench("buffer_create_empty", ITERATIONS, lambda: Buffer(0)))
    results.append(bench("buffer_create_1KB", ITERATIONS, lambda: Buffer(1024)))

    # Buffer from data
    data_256 = bytes(256)
    results.append(
        bench("buffer_from_slice_256B", ITERATIONS, lambda: Buffer.from_slice(data_256))
    )

    data_1k = bytes(1024)
    results.append(
        bench("buffer_from_slice_1KB", ITERATIONS, lambda: Buffer.from_slice(data_1k))
    )

    # Buffer append
    chunk = bytes(64)

    def buffer_append_64b() -> None:
        buf = Buffer(64)
        buf.append_data(chunk)

    results.append(bench("buffer_append_64B", ITERATIONS // 10, buffer_append_64b))

    print()

    # ========================================================================
    # Combined Operations
    # ========================================================================

    print("--- Combined Operations ---\n")

    # Simulate page rendering setup
    def page_render_setup() -> None:
        dpi = 144.0
        scale = dpi / 72.0
        ctm = Matrix.scale(scale, scale)
        page_bounds = Rect(0.0, 0.0, 612.0, 792.0)
        page_bounds.transform(ctm)

    results.append(bench("page_render_setup", ITERATIONS // 10, page_render_setup))

    # Simulate text position calculation
    def text_position_calc() -> None:
        base = Point(72.0, 700.0)
        tm = Matrix.translate(0.0, -14.0)
        base.transform(tm)

    results.append(bench("text_position_calc", ITERATIONS, text_position_calc))

    # Simulate bounding box calculation for 10 items
    def bbox_calc_10_items() -> None:
        bbox = Rect(1e30, 1e30, -1e30, -1e30)
        for i in range(10):
            item = Rect(i * 10, i * 10, 100 + i * 10, 20 + i * 10)
            bbox = bbox.union(item)

    results.append(bench("bbox_calc_10_items", ITERATIONS // 10, bbox_calc_10_items))

    print()

    # ========================================================================
    # Summary
    # ========================================================================

    print("=== Summary ===\n")

    for result in results:
        result.print()

    print()

    # Output JSON for cross-language comparison
    print("--- JSON Output ---")
    print(json.dumps([r.to_dict() for r in results], indent=2))


if __name__ == "__main__":
    main()

